distributed/comm/ucx.py

```diff
@@ -0,0 +1,366 @@
+        nframes, = struct.unpack("Q", obj[:8])  # first eight bytes for number of frames
+
+        gpu_frame_msg = obj[
+            8 : 8 + nframes
+        ]  # next nframes bytes for if they're GPU frames
+        is_gpus = struct.unpack("{}?".format(nframes), gpu_frame_msg)
+
+        sized_frame_msg = obj[8 + nframes :]  # then the rest for frame sizes
+        sizes = struct.unpack("{}Q".format(nframes), sized_frame_msg)
+
+        frames = []
+
+        for i, (is_gpus, size) in enumerate(zip(is_gpus, sizes)):
+            if size > 0:
+                resp = await self.ep.recv_obj(size, cuda=is_gpus)
+            else:
+                resp = await self.ep.recv_future()
+            frame = ucp.get_obj_from_msg(resp)
+            frames.append(frame)
+
+        msg = await from_frames(
+            frames, deserialize=self.deserialize, deserializers=deserializers
+        )
+
+        return msg
+
+    def abort(self):
+        # breakpoint()
+        if self._ep:
+            ucp.destroy_ep(self._ep)
+            logger.debug("Destroyed UCX endpoint")
+            self._ep = None
+        # if self.listener_instance:
+        # ucp.stop_listener(self.listener_instance)
+        # self.listener_instance = None
+
+    @property
+    def ep(self):
+        if self._ep:
+            return self._ep
+        else:
+            raise CommClosedError("UCX Endpoint is closed")
+
+    async def close(self):
+        # TODO: Handle in-flight messages?
+        # sleep is currently used to help flush buffer
+        self.abort()
+
+    def closed(self):
+        return self._ep is None
+
+
+class UCXConnector(Connector):
+    prefix = "ucx://"
+    comm_class = UCX
+    encrypted = False
+
+    client = ...  # TODO: add a client here?
+
+    async def connect(self, address: str, deserialize=True, **connection_args) -> UCX:
+        logger.debug("UCXConnector.connect: %s", address)
+        _ucp_init()
+        ip, port = _parse_host_port(address)
+        ep = await ucp.get_endpoint(ip.encode(), port)
+        return self.comm_class(
+            ep, self.prefix + address, listener_instance=None, deserialize=deserialize
+        )
+
+
+class UCXListener(Listener):
+    # MAX_LISTENERS 256 in ucx-py
+    prefix = UCXConnector.prefix
+    comm_class = UCXConnector.comm_class
+    encrypted = UCXConnector.encrypted
+
+    def __init__(
+        self, address: str, comm_handler: None, deserialize=False, **connection_args
+    ):
+        logger.debug("UCXListener.__init__")
+        if not address.startswith("ucx"):
+            address = "ucx://" + address
+        self.address = address
+        self.ip, self.port = _parse_host_port(address, default_port=0)
+        self.comm_handler = comm_handler
+        self.deserialize = deserialize
+        self._ep = None  # type: ucp.Endpoint
+        self.listener_instance = None  # type: ucp.ListenerFuture
+        self._task = None
+
+        # XXX: The init may be required to take args like
```

Comment:
I think this was a comment about the abc `Listener.__init__`, but I'm not sure anymore.